### ISSUE TITLE
gss_add_cred() crashes on null output_cred_handle

### ISSUE THREAD

**Issue by greghudson** [heimdal/heimdal#413]
RFC 2744 says about gss_add_cred():
```
   This routine can be used to either compose a new credential
   containing all credential-elements of the original in addition to the
   newly-acquire credential-element, or to add the new credential-
   element to an existing credential. If NULL is specified for the
   output_cred_handle parameter argument, the new credential-element
   will be added to the credential identified by input_cred_handle; if a
   valid pointer is specified for the output_cred_handle parameter, a
   new credential handle will be created.
```
Heimdal's mechglue gss_add_cred() does not expect output_cred_handle to be NULL; it dereferences it unconditionally at the beginning and end of the function, and only contains logic to create a new credential (with or without a copy of mechanism creds from an old credential), not to mutate an existing one.

(I discovered this through code inspection while looking into bugs in MIT krb5's version of the same function.  I have not tried to reproduce it, and I don't believe gss_add_cred() is used much in the wild.)

**Comment by lhoward** [comment]
Should be a trivial fix

**Comment by lhoward** [comment]
should `input_cred_handle` really be `gss_const_cred_id_t` if it can be mutated? or it just a const pointer?

**Comment by lhoward** [comment]
Seems like this really needs to be rewritten to the spec

**Comment by nicowilliams** [comment]
~input_cred_handle` should not be `gss_const_cred_id_t`, indeed.  The spec is somewhat confusing as to `gss_add_cred()`, and we've had debates as to its meaning, though I think the intent is pretty clear.

**Comment by greghudson** [comment]
From a language perspective, that `const` doesn't do anything to the function contract.  It merely prevents the function definition from modifying the variable containing the argument (but it can still freely modify what that variable points to).  This is a common botch in RFC 2744.  This particular instance of it is a double botch which cancels itself out--if it had the intended effect, it would contradict the text, but since it doesn't have the intended effect, it's harmless.

**Comment by nicowilliams** [comment]
Yeah, I know.  C really needs an `immutable`, but it doesn't have one.

**Comment by nicowilliams** [comment]
Heimdal's `gss_add_cred()` does need almost a complete re-write :(

**Comment by nicowilliams** [comment]
Oh, and Viktor and I have discussed this many times before.  I've been resisting spending the effort on this.  I'll take a look.

**Comment by lhoward** [comment]
It's non-trivial but it's probably not too bad. But maybe best done with `gss_add_cred_from()` implementation?

**Comment by nicowilliams** [comment]
I've got a first attempt ready.  Gotta write some tests top.

**Comment by lhoward** [comment]
is `gss_duplicate_cred()` something that pluggable mechanisms should be able to implement? if so you need to add a `OPTSYM` entry to `gss_mech_switch.c`.

**Comment by nicowilliams** [comment]
Oh, right, I'll fix that!
